### PR TITLE
feat(DockerMock): calls should show up in the call stack

### LINKED ISSUE #406: DockerMock calls don't show up in the call stack
### Feature Request

Most of the time when we write unit tests, we're verifying the callStack like in the examples:

```groovy
@Test
void should_execute_without_errors() throws Exception {
    runScript("Jenkinsfile")
    assertThat(helper.callStack.findAll { call ->
        call.methodName == "sh"
    }.any { call ->
        callArgsToString(call).contains("mvn verify")
    }).isTrue()
    assertJobStatusSuccess()
}
```

But when we use the `docker-workflow-plugin` all the calls to e.g. `docker.image` or `docker.build` aren't added,
when we add a binding like `binding.setVariable('docker', new DockerMock())`.

One way to fix this, is to modify the call-stack like in `PipelineTestHelper.groovy`, but I don't know if that's an inelegant solution, and it's easier to do in another way?

I've included an example below of what it would look like for e.g. the `image`-method:

```groovy
package com.lesfurets.jenkins.unit

@SuppressWarnings(['EmptyMethod', 'MethodReturnTypeRequired', 'UnusedMethodParameter'])
class DockerMock implements Serializable {

    PipelineTestHelper helper
    Object script

    DockerMock(script = null, helper = null){
        this.script = script
        this.helper = helper
    }

    ...

    Image image(String id) {
        if (script != null && helper != null){
            int depth = Thread.currentThread().stackTrace.findAll { it.className == delegate.class.name }.size()
            helper.registerMethodCall(script, depth, "docker.image", id)
        }
        return new Image(id)
    }
}
```

The null-checks are there so it can be used along with just `DockerMock()`.

The test script has access to the `helper` since it runs `helper.callStack.findAll..`. But it will need access to the `script` with a 
```groovy
script = loadScript("Jenkinsfile")
```
And not just `runScript("Jenkinsfile")` like in the example. But an example could be provided for that. Usage is then like,

```groovy
script = loadScript("Jenkinsfile")
binding.setVariable('docker', new DockerMock(script, helper))
```

--- 

Currently we have our own `ExtendedDockerMock.groovy` inspired by <https://github.com/jenkinsci/JenkinsPipelineUnit/issues/362#issuecomment-859482509> which looks like..

```groovy
import com.lesfurets.jenkins.unit.DockerMock
import com.lesfurets.jenkins.unit.PipelineTestHelper

@SuppressWarnings(['EmptyMethod', 'MethodReturnTypeRequired', 'UnusedMethodParameter'])
class ExtendedDockerMock extends DockerMock {

    PipelineTestHelper helper
    Object script

    ExtendedDockerMock(script, helper){
        super()
        this.script = script
        this.helper = helper
    }

    Image image(String id) {
        Image image = super.image(id)

        int depth = Thread.currentThread().stackTrace.findAll { it.className == delegate.class.name }.size()
        helper.registerMethodCall(script, depth, "docker.image", id)

        return image
    }

}
```

But ideally we'd just get a solution in `DockerMock`.

The entire thing is a bit reverse-engineered from the `PipelineTestHelper`. But before spending a lot of time creating a PR, I'd like to know if this is something you want? Another alternative is the one mentioned in https://github.com/jenkinsci/JenkinsPipelineUnit/issues/362. it doesn't require the `script` or `helper` but you have to inspect the `DockerMock` directly.

### PR DESCRIPTION
**Pull request by stanovchar**
Refactored Docker Mock tests. Setup Interceptors to trace docker mock methods in the call stack.

Fixes #406 

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by stchar** [comment]
Oh grep, just noticed that I had used another account to open the PR.

**Comment by nre-ableton** [comment]
BTW, I'm ok with this PR, but of course it'll need to be reworked to drop c1dd159 and remove the multiple arguments in some of the tests.

**Comment by stchar** [comment]
Rebased. No more dependency on #534

**Review by nre-ableton — CHANGES_REQUESTED** [review]

File: src/test/jenkins/jenkinsfiles/Docker_Jenkinsfile [hunk 1]
```diff
@@ -1,76 +1,107 @@
...
 
-    stage('Test docker.withRegistry') {
-      docker.withRegistry('test.registry', 'fake-credentials') {
-        def image = docker.image(testImageName)
-        assert image.id == testImageName
-      }
-    }
+def test_docker_build_multiple_args(testImageName) {
+  stage('Test docker.build() multiple args') {
+    def image = docker.build(testImageName, '--build_arg arg1=val1', '--build_arg arg2=val2' )
```
[the remark this answers is not in the document]
Comment (stchar) [reply src/test/jenkins/jenkinsfiles/Docker_Jenkinsfile]: yep, removed

**Review by nre-ableton — APPROVED** [review]
Sorry, please disregard my previous review, I submitted it just seconds before you force-pushed. :smile: Anyways, LGTM, assuming that the tests pass.

**Pull request merged**